ajv/src/ajv.ts

```diff
@@ -0,0 +1,112 @@
+import { toNestError, validateFieldsNatively } from '@hookform/resolvers';
+import Ajv, { DefinedError, KeywordCxt, _ as genCode } from 'ajv';
+import ajvErrors from 'ajv-errors';
+import { appendErrors, FieldError } from 'react-hook-form';
+import { Resolver } from './types';
+
+const parseErrorSchema = (
+  ajvErrors: DefinedError[],
+  validateAllFieldCriteria: boolean,
+) => {
+  // Ajv will return empty instancePath when require error
+  ajvErrors.forEach((error) => {
+    if (error.keyword === 'required') {
+      error.instancePath = '/' + error.params.missingProperty;
+    }
+  });
+
+  return ajvErrors.reduce<Record<string, FieldError>>((previous, error) => {
+    // `/deepObject/data` -> `deepObject.data`
+    const path = error.instancePath.substring(1).replace('/', '.');
+
+    if (!previous[path]) {
+      previous[path] = {
+        message: error.message,
+        type: error.keyword,
+      };
+    }
+
+    if (validateAllFieldCriteria) {
+      const types = previous[path].types;
+      const messages = types && types[error.keyword];
+
+      previous[path] = appendErrors(
+        path,
+        validateAllFieldCriteria,
+        previous,
+        error.keyword,
+        messages
+          ? ([] as string[]).concat(messages as string[], error.message || '')
+          : error.message,
+      ) as FieldError;
+    }
+
+    return previous;
+  }, {});
+};
+
+export const ajvResolver: Resolver =
+  (schema, schemaOptions, resolverOptions = {}) =>
+  async (values, _, options) => {
+    const ajv = new Ajv({
+      allErrors: true,
+      validateSchema: true,
+      ...schemaOptions,
+    });
+
+    ajvErrors(ajv);
+
+    // If the field is not require in form, the value is empty (value: "")
+    // But Ajv `required` cannot check empty, it can only check undefined
+    // This keyword is for checking empty, but not available for nest fields
+    // https://stackoverflow.com/a/72374873/15246747
+    ajv.addKeyword({
```

Comment:
I've solved the seconed issue. But if the field is nest (like deep object, array, etc.), it may fail.

Reply:
Also, this only support string value. I think it it unnecessary to implement this. We can add some tips in the documentation if user meet required error.